File: crates/less-sync-core/src/reencrypt.rs
//! DEK re-wrapping and epoch forward derivation.

use crate::error::SyncError;
use less_crypto::{derive_next_epoch_key, unwrap_dek, wrap_dek};
use std::collections::HashMap;
use zeroize::Zeroize;

/// Read the epoch prefix from a wrapped DEK (first 4 bytes, big-endian u32).
pub fn peek_epoch(wrapped_dek: &[u8]) -> Result<u32, SyncError> {
    if wrapped_dek.len() < 4 {
        return Err(SyncError::MissingDek);
    }
    Ok(u32::from_be_bytes(
        wrapped_dek[..4]
            .try_into()
            .expect("4 bytes after length check"),
    ))
}

/// Derive a key forward from one epoch to another by chaining `derive_next_epoch_key`.
///
/// # Arguments
/// * `key` - Starting epoch key (32 bytes)
/// * `space_id` - Space ID for domain separation
/// * `from_epoch` - Starting epoch number
/// * `to_epoch` - Target epoch number (must be >= from_epoch)
pub fn derive_forward(
    key: &[u8],
    space_id: &str,
    from_epoch: u32,
    to_epoch: u32,
) -> Result<Vec<u8>, SyncError> {
    if to_epoch < from_epoch {
        return Err(SyncError::BackwardDerivation {
            target: to_epoch,
            base: from_epoch,
        });
    }
    if to_epoch == from_epoch {
        return Ok(key.to_vec());
    }
    let mut current = key.to_vec();
    for e in (from_epoch + 1)..=to_epoch {
        current = derive_next_epoch_key(&current, space_id, e)?.to_vec();
    }
    Ok(current)
}

/// Re-wrap a set of DEKs from their current epoch to a new epoch key.
///
/// Builds a key cache from `current_epoch` to `new_epoch` to handle DEKs
/// at any intermediate epoch. Returns the re-wrapped DEK bytes.
///
/// # Arguments
/// * `wrapped_deks` - Pairs of (id, wrapped_dek_bytes)
/// * `current_key` - Current epoch key (32 bytes)
/// * `current_epoch` - Current epoch number
/// * `new_key` - Target epoch key (32 bytes)
/// * `new_epoch` - Target epoch number
/// * `space_id` - Space ID for domain separation
pub fn rewrap_deks(
    wrapped_deks: &[(String, Vec<u8>)],
    current_key: &[u8],
    current_epoch: u32,
    new_key: &[u8],
    new_epoch: u32,
    space_id: &str,
) -> Result<Vec<(String, Vec<u8>)>, SyncError> {
    if new_epoch <= current_epoch {
        return Err(SyncError::InvalidEpochAdvance {
            new: new_epoch,
            current: current_epoch,
        });
    }

    // Build key cache for unwrapping DEKs at any epoch in [current_epoch, new_epoch]
    let mut key_cache = HashMap::new();
    key_cache.insert(current_epoch, current_key.to_vec());
    let mut derived_key = current_key.to_vec();
    for e in (current_epoch + 1)..=new_epoch {
        derived_key = derive_next_epoch_key(&derived_key, space_id, e)?.to_vec();
        key_cache.insert(e, derived_key.clone());
    }
    derived_key.zeroize();

    let mut result = Vec::new();
    for (id, wrapped_dek) in wrapped_deks {
        let dek_epoch = peek_epoch(wrapped_dek)?;
        if dek_epoch == new_epoch {
            continue; // Already at target epoch
        }

        let unwrap_key = key_cache
            .get(&dek_epoch)
            .ok_or(SyncError::NoKek(dek_epoch))?;

        let (mut dek, _epoch) = unwrap_dek(wrapped_dek, unwrap_key)?;
        let rewrapped = wrap_dek(&dek, new_key, new_epoch)?;
        dek.zeroize();

        result.push((id.clone(), rewrapped.to_vec()));
    }

    // Zero derived intermediate keys (not current_key — caller owns it)
    for (epoch, mut key) in key_cache {
        if epoch != current_epoch {
            key.zeroize();
        }
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use less_crypto::{generate_dek, wrap_dek as crypto_wrap_dek};

    fn random_key() -> [u8; 32] {
        let mut key = [0u8; 32];
        getrandom::getrandom(&mut key).unwrap();
        key
    }

    #[test]
    fn peek_epoch_reads_big_endian() {
        let mut data = vec![0u8; 44];
        data[0] = 0x00;
        data[1] = 0x00;
        data[2] = 0x00;
        data[3] = 0x05;
        assert_eq!(peek_epoch(&data).unwrap(), 5);
    }

    #[test]
    fn peek_epoch_rejects_short() {
        assert!(peek_epoch(&[1, 2, 3]).is_err());
    }

    #[test]
    fn derive_forward_same_epoch() {
        let key = random_key();
        let result = derive_forward(&key, "space-1", 0, 0).unwrap();
        assert_eq!(result, key);
    }

    #[test]
    fn derive_forward_multiple_steps() {
        let key = random_key();
        let result = derive_forward(&key, "space-1", 0, 3).unwrap();
        assert_ne!(result, key.to_vec());
        assert_eq!(result.len(), 32);
    }

    #[test]
    fn derive_forward_rejects_backward() {
        let key = random_key();
        assert!(derive_forward(&key, "space-1", 5, 3).is_err());
    }

    #[test]
    fn rewrap_deks_round_trip() {
        let key1 = random_key();
        let space_id = "space-1";

        // Create some DEKs wrapped at epoch 1
        let dek1 = generate_dek();
        let dek2 = generate_dek();
        let wrapped1 = crypto_wrap_dek(&dek1, &key1, 1).unwrap();
        let wrapped2 = crypto_wrap_dek(&dek2, &key1, 1).unwrap();

        let wrapped_deks = vec![
            ("rec-1".to_string(), wrapped1.to_vec()),
            ("rec-2".to_string(), wrapped2.to_vec()),
        ];

        // Derive key for epoch 2
        let key2 = derive_next_epoch_key(&key1, space_id, 2).unwrap();

        // Rewrap from epoch 1 to epoch 2
        let rewrapped = rewrap_deks(&wrapped_deks, &key1, 1, &key2, 2, space_id).unwrap();

        assert_eq!(rewrapped.len(), 2);

        // Verify epoch prefix is updated
        assert_eq!(peek_epoch(&rewrapped[0].1).unwrap(), 2);
        assert_eq!(peek_epoch(&rewrapped[1].1).unwrap(), 2);

        // Verify DEKs can be unwrapped with new key
        let (unwrapped1, _) = unwrap_dek(&rewrapped[0].1, &key2).unwrap();
        let (unwrapped2, _) = unwrap_dek(&rewrapped[1].1, &key2).unwrap();
        assert_eq!(unwrapped1, dek1);
        assert_eq!(unwrapped2, dek2);
    }

    #[test]
    fn rewrap_mixed_epoch_deks() {
        let key1 = random_key();
        let space_id = "space-1";

        // DEK at epoch 1
        let dek_a = generate_dek();
        let wrapped_a = crypto_wrap_dek(&dek_a, &key1, 1).unwrap();

        // DEK at epoch 2
        let key2 = derive_next_epoch_key(&key1, space_id, 2).unwrap();
        let dek_b = generate_dek();
        let wrapped_b = crypto_wrap_dek(&dek_b, &key2, 2).unwrap();

        let wrapped_deks = vec![
            ("rec-a".to_string(), wrapped_a.to_vec()),
            ("rec-b".to_string(), wrapped_b.to_vec()),
        ];

        // Rewrap to epoch 3
        let key3 = derive_next_epoch_key(&key2, space_id, 3).unwrap();
        let rewrapped = rewrap_deks(&wrapped_deks, &key1, 1, &key3, 3, space_id).unwrap();

        assert_eq!(rewrapped.len(), 2);
        for (_, w) in &rewrapped {
            assert_eq!(peek_epoch(w).unwrap(), 3);
        }

        // Verify original DEKs are recoverable
        let (unwrapped_a, _) = unwrap_dek(&rewrapped[0].1, &key3).unwrap();
        let (unwrapped_b, _) = unwrap_dek(&rewrapped[1].1, &key3).unwrap();
        assert_eq!(unwrapped_a, dek_a);
        assert_eq!(unwrapped_b, dek_b);
    }

    #[test]
    fn rewrap_skips_already_at_target() {
        let key = random_key();
        let space_id = "space-1";

        let dek = generate_dek();
        let wrapped = crypto_wrap_dek(&dek, &key, 2).unwrap(); // Already at epoch 2

        let key2 = derive_next_epoch_key(&key, space_id, 2).unwrap();
        let result = rewrap_deks(
            &[("rec-1".to_string(), wrapped.to_vec())],
            &key,
            1,
            &key2,
            2,
            space_id,
        )
        .unwrap();

        assert!(result.is_empty()); // Skipped
    }
}

File: js/src/sync/reencrypt.ts
/**
 * DEK re-wrapping for epoch advancement (forward secrecy).
 *
 * With per-record DEKs, epoch rotation no longer requires re-encrypting
 * record blobs. Instead:
 * 1. epoch.begin RPC to advance server (CAS)
 * 2. Fetch all wrapped DEKs (lightweight — no blobs)
 * 3. Unwrap each DEK with old KEK, re-wrap with new KEK
 * 4. Upload re-wrapped DEKs
 * 5. epoch.complete RPC to clear rewrap_epoch
 */

import { deriveNextEpochKey } from "../crypto/index.js";
import { unwrapDEK, wrapDEK } from "../crypto/internals.js";
import {
  webcryptoWrapDEK,
  webcryptoUnwrapDEK,
  webcryptoDeriveEpochKey,
} from "../crypto/webcrypto.js";
import type { WSClient } from "./ws-client.js";
import type { WSEpochConflictResult } from "./ws-frames.js";
import { RPCCallError } from "./rpc-connection.js";

/**
 * Thrown when epoch advancement fails due to server state mismatch (conflict).
 * Contains the server's current state for the caller to handle.
 */
export class EpochMismatchError extends Error {
  constructor(
    public currentEpoch: number,
    public rewrapEpoch: number | null,
  ) {
    super(
      `Epoch mismatch: server at epoch ${currentEpoch}` +
        (rewrapEpoch !== null ? `, rewrap_epoch=${rewrapEpoch}` : ""),
    );
    this.name = "EpochMismatchError";
  }
}

/** Configuration for advanceEpoch. */
export interface AdvanceEpochConfig {
  ws: WSClient;
  spaceId: string;
  ucan?: string;
}

/** Options for advanceEpoch. */
export interface AdvanceEpochOptions {
  /** Set min_epoch to the new epoch (for revocation — skips grace period). */
  setMinEpoch?: boolean;
}

/**
 * Advance the epoch on the server (CAS operation).
 *
 * Sets `rewrap_epoch` atomically. The caller must rewrap DEKs and call
 * epochComplete to clear the flag.
 *
 * @throws EpochMismatchError on conflict with structured server state
 * @throws Error on other failures
 */
export async function advanceEpoch(
  config: AdvanceEpochConfig,
  newEpoch: number,
  opts?: AdvanceEpochOptions,
): Promise<void> {
  const result = await config.ws.epochBegin({
    space: config.spaceId,
    ...(config.ucan ? { ucan: config.ucan } : {}),
    epoch: newEpoch,
    ...(opts?.setMinEpoch ? { set_min_epoch: true } : {}),
  });

  // Check for conflict result (returned as success with error field).
  // The server sends error code "conflict" (ERR_CODE_CONFLICT); tolerate the
  // legacy "epoch_conflict" spelling as well (AUD-030: previously only
  // "epoch_conflict" was recognized, so real conflicts were treated as
  // success and rotation recovery never ran).
  if (
    "error" in result &&
    ((result as WSEpochConflictResult).error === "conflict" ||
      (result as WSEpochConflictResult).error === "epoch_conflict")
  ) {
    const conflict = result as WSEpochConflictResult;
    throw new EpochMismatchError(
      conflict.current_epoch,
      conflict.rewrap_epoch ?? null,
    );
  }
}

/** Configuration for rewrapAllDEKs. */
export interface RewrapAllDEKsConfig {
  /** WSClient for DEK RPC operations. */
  ws: WSClient;
  /** Space ID (for key derivation domain separation). */
  spaceId: string;
  /** Optional UCAN for shared space authorization. */
  ucan?: string;
  /** Current epoch number. */
  currentEpoch: number;
  /** Current epoch key — raw bytes (shared spaces) or CryptoKey (personal space). */
  currentKey: Uint8Array | CryptoKey;
  /** Current HKDF derive key (for CryptoKey path). */
  currentDeriveKey?: CryptoKey;
  /** Target epoch number. */
  newEpoch: number;
  /** Target epoch key — raw bytes (shared spaces) or CryptoKey (personal space). */
  newKey: Uint8Array | CryptoKey;
  /** Whether to also rewrap file DEKs (default: true). */
  includeFiles?: boolean;
  /**
   * Fresh-key rotation (AUD-024): newKey is a random secret, NOT derived
   * from currentKey. The key cache then holds exactly {currentEpoch,
   * currentKey} and {newEpoch, newKey} — no forward derivation.
   */
  freshKey?: boolean;
}

export interface RewrapResult {
  /** Number of record DEKs re-wrapped. */
  dekCount: number;
  /** Number of file DEKs re-wrapped. */
  fileDekCount: number;
}

/**
 * Re-wrap all DEKs in a space under a new epoch key.
 *
 * Fetches all wrapped DEKs, unwraps with the appropriate epoch key,
 * re-wraps with the new key, and uploads. Idempotent — skips DEKs
 * already at the target epoch.
 */
export async function rewrapAllDEKs(
  config: RewrapAllDEKsConfig,
): Promise<RewrapResult> {
  const { ws, spaceId, ucan, currentEpoch, currentKey, newEpoch, newKey } =
    config;
  const includeFiles = config.includeFiles !== false;

  if (newEpoch <= currentEpoch) {
    throw new Error(
      `Invalid epoch: newEpoch=${newEpoch} must be greater than currentEpoch=${currentEpoch}`,
    );
  }

  // CryptoKey path: use Web Crypto for unwrap/wrap
  if (currentKey instanceof CryptoKey) {
    return rewrapAllDEKsCryptoKey(
      config as RewrapAllDEKsConfig & {
        currentKey: CryptoKey;
        newKey: CryptoKey;
      },
    );
  }

  // Raw bytes path: use WASM
  const rawNewKey = newKey as Uint8Array;

  // Build key cache for unwrapping DEKs. Legacy rotation derives the chain
  // forward; fresh-key rotation (AUD-024) holds exactly the old and new keys.
  const keyCache = new Map<number, Uint8Array>();
  keyCache.set(currentEpoch, currentKey);
  if (config.freshKey) {
    keyCache.set(newEpoch, rawNewKey);
  } else {
    let derivedKey: Uint8Array = currentKey;
    for (let e = currentEpoch + 1; e <= newEpoch; e++) {
      derivedKey = deriveNextEpochKey(derivedKey, spaceId, e);
      keyCache.set(e, derivedKey);
    }
  }

  try {
    // Re-wrap record DEKs. Each entry carries the wrapper we observed so the
    // server applies a compare-and-set; a concurrent push replaces a wrapper,
    // and rewrapping from a stale read would make that record undecryptable
    // (AUD-026). On conflict the whole pass refetches and retries.
    const rewrapped: Array<{
      id: string;
      wrapped_dek: Uint8Array;
      observed_wrapped_dek: Uint8Array;
    }> = [];
    for (let attempt = 0; ; attempt++) {
      const deks = await ws.getDEKs({
        space: spaceId,
        ...(ucan ? { ucan } : {}),
        since: 0,
      });
      rewrapped.length = 0;
      for (const { id, wrapped_dek: wrappedDEK } of deks) {
        const dekEpoch = peekEpoch(wrappedDEK);
        if (dekEpoch === newEpoch) continue;
        const unwrapKey = keyCache.get(dekEpoch);
        if (!unwrapKey) throw new Error(`No key for DEK epoch ${dekEpoch}`);
        const { dek } = unwrapDEK(wrappedDEK, unwrapKey);
        try {
          rewrapped.push({
            id,
            wrapped_dek: wrapDEK(dek, rawNewKey, newEpoch),
            observed_wrapped_dek: wrappedDEK,
          });
        } finally {
          dek.fill(0);
        }
      }
      if (rewrapped.length === 0) break;
      try {
        const result = await ws.rewrapDEKs({
          space: spaceId,
          ...(ucan ? { ucan } : {}),
          deks: rewrapped,
        });
        if (!result.ok) throw new Error("DEK re-wrapping failed on server");
        break;
      } catch (err) {
        if (!isRetryableRewrapConflict(err) || attempt >= REWRAP_MAX_RETRIES) {
          throw err;
        }
      }
    }

    // Re-wrap file DEKs (same compare-and-set discipline)
    let fileDekCount = 0;
    if (includeFiles) {
      const rewrappedFiles: Array<{
        id: string;
        wrapped_dek: Uint8Array;
        observed_wrapped_dek: Uint8Array;
      }> = [];
      for (let attempt = 0; ; attempt++) {
        const fileDeks = await ws.getFileDEKs({
          space: spaceId,
          ...(ucan ? { ucan } : {}),
          since: 0,
        });
        rewrappedFiles.length = 0;
        for (const { id, wrapped_dek: wrappedDEK } of fileDeks) {
          const dekEpoch = peekEpoch(wrappedDEK);
          if (dekEpoch === newEpoch) continue;
          const unwrapKey = keyCache.get(dekEpoch);
          if (!unwrapKey)
            throw new Error(`No key for file DEK epoch ${dekEpoch}`);
          const { dek } = unwrapDEK(wrappedDEK, unwrapKey);
          try {
            rewrappedFiles.push({
              id,
              wrapped_dek: wrapDEK(dek, rawNewKey, newEpoch),
              observed_wrapped_dek: wrappedDEK,
            });
          } finally {
            dek.fill(0);
          }
        }
        if (rewrappedFiles.length === 0) break;
        try {
          const result = await ws.rewrapFileDEKs({
            space: spaceId,
            ...(ucan ? { ucan } : {}),
            deks: rewrappedFiles,
          });
          if (!result.ok)
            throw new Error("File DEK re-wrapping failed on server");
          break;
        } catch (err) {
          if (
            !isRetryableRewrapConflict(err) ||
            attempt >= REWRAP_MAX_RETRIES
          ) {
            throw err;
          }
        }
      }
      fileDekCount = rewrappedFiles.length;
    }

    return { dekCount: rewrapped.length, fileDekCount };
  } finally {
    // Zero derived intermediate keys only — the caller owns currentKey and,
    // for fresh-key rotation, newKey as well.
    for (const [epoch, key] of keyCache) {
      if (epoch !== currentEpoch && !(config.freshKey && epoch === newEpoch)) {
        key.fill(0);
      }
    }
    keyCache.clear();
  }
}

/** Maximum epoch gap for forward derivation (defense-in-depth). */
const MAX_EPOCH_ADVANCE = 1000;

/** Bounded refetch+retry passes when a rewrap loses a compare-and-set race. */
const REWRAP_MAX_RETRIES = 3;

/** A rewrap conflict (concurrent replacement or epoch mismatch) is retryable. */
function isRetryableRewrapConflict(err: unknown): boolean {
  return err instanceof RPCCallError && err.code === "conflict";
}

/**
 * CryptoKey path for rewrapAllDEKs — uses Web Crypto for unwrap/wrap.
 */
async function rewrapAllDEKsCryptoKey(
  config: RewrapAllDEKsConfig & { currentKey: CryptoKey; newKey: CryptoKey },
): Promise<RewrapResult> {
  const {
    ws,
    spaceId,
    ucan,
    currentEpoch,
    currentKey,
    currentDeriveKey,
    newEpoch,
    newKey,
  } = config;
  const includeFiles = config.includeFiles !== false;

  const distance = newEpoch - currentEpoch;
  if (distance > MAX_EPOCH_ADVANCE) {
    throw new Error(
      `Epoch gap too large: ${distance} (max: ${MAX_EPOCH_ADVANCE}). ` +
        `This may indicate corrupted state or a malicious server.`,
    );
  }

  // Build CryptoKey cache for unwrapping DEKs at any epoch in [currentEpoch, newEpoch]
  const kwKeyCache = new Map<number, CryptoKey>();
  kwKeyCache.set(currentEpoch, currentKey);

  if (currentDeriveKey && newEpoch > currentEpoch) {
    let dk = currentDeriveKey;
    for (let e = currentEpoch + 1; e <= newEpoch; e++) {
      const derived = await webcryptoDeriveEpochKey(dk, spaceId, e);
      kwKeyCache.set(e, derived.kwKey);
      dk = derived.deriveKey;
    }
  }

  async function rewrapDEKList(
    deks: Array<{ id: string; wrapped_dek: Uint8Array }>,
  ): Promise<
    Array<{
      id: string;
      wrapped_dek: Uint8Array;
      observed_wrapped_dek: Uint8Array;
    }>
  > {
    const rewrapped: Array<{
      id: string;
      wrapped_dek: Uint8Array;
      observed_wrapped_dek: Uint8Array;
    }> = [];
    for (const { id, wrapped_dek: wrappedDEK } of deks) {
      const dekEpoch = peekEpoch(wrappedDEK);
      if (dekEpoch === newEpoch) continue;
      const unwrapKey = kwKeyCache.get(dekEpoch);
      if (!unwrapKey) throw new Error(`No key for DEK epoch ${dekEpoch}`);
      const { dek } = await webcryptoUnwrapDEK(wrappedDEK, unwrapKey);
      try {
        rewrapped.push({
          id,
          wrapped_dek: await webcryptoWrapDEK(dek, newKey, newEpoch),
          observed_wrapped_dek: wrappedDEK,
        });
      } finally {
        dek.fill(0);
      }
    }
    return rewrapped;
  }

  try {
    // Re-wrap record DEKs with the same compare-and-set discipline as the
    // raw-key path: each entry carries the observed wrapper, and a conflict
    // refetches and retries instead of clobbering (AUD-026).
    const rewrapped: Array<{
      id: string;
      wrapped_dek: Uint8Array;
      observed_wrapped_dek: Uint8Array;
    }> = [];
    for (let attempt = 0; ; attempt++) {
      const deks = await ws.getDEKs({
        space: spaceId,
        ...(ucan ? { ucan } : {}),
        since: 0,
      });
      rewrapped.length = 0;
      rewrapped.push(...(await rewrapDEKList(deks)));
      if (rewrapped.length === 0) break;
      try {
        const result = await ws.rewrapDEKs({
          space: spaceId,
          ...(ucan ? { ucan } : {}),
          deks: rewrapped,
        });
        if (!result.ok) throw new Error("DEK re-wrapping failed on server");
        break;
      } catch (err) {
        if (!isRetryableRewrapConflict(err) || attempt >= REWRAP_MAX_RETRIES) {
          throw err;
        }
      }
    }

    // Re-wrap file DEKs (same compare-and-set discipline)
    let fileDekCount = 0;
    if (includeFiles) {
      const rewrappedFiles: Array<{
        id: string;
        wrapped_dek: Uint8Array;
        observed_wrapped_dek: Uint8Array;
      }> = [];
      for (let attempt = 0; ; attempt++) {
        const fileDeks = await ws.getFileDEKs({
          space: spaceId,
          ...(ucan ? { ucan } : {}),
          since: 0,
        });
        rewrappedFiles.length = 0;
        rewrappedFiles.push(...(await rewrapDEKList(fileDeks)));
        if (rewrappedFiles.length === 0) break;
        try {
          const result = await ws.rewrapFileDEKs({
            space: spaceId,
            ...(ucan ? { ucan } : {}),
            deks: rewrappedFiles,
          });
          if (!result.ok)
            throw new Error("File DEK re-wrapping failed on server");
          break;
        } catch (err) {
          if (
            !isRetryableRewrapConflict(err) ||
            attempt >= REWRAP_MAX_RETRIES
          ) {
            throw err;
          }
        }
      }
      fileDekCount = rewrappedFiles.length;
    }

    return { dekCount: rewrapped.length, fileDekCount };
  } finally {
    // Drop CryptoKey references so they can be garbage collected
    kwKeyCache.clear();
  }
}

/** Read the epoch prefix from a wrapped DEK without unwrapping it (first 4 bytes, big-endian u32). */
export function peekEpoch(wrappedDEK: Uint8Array): number {
  return new DataView(
    wrappedDEK.buffer,
    wrappedDEK.byteOffset,
    wrappedDEK.byteLength,
  ).getUint32(0, false);
}

/**
 * Derive a key forward from one epoch to another by chaining deriveNextEpochKey.
 */
export function deriveForward(
  key: Uint8Array,
  spaceId: string,
  fromEpoch: number,
  toEpoch: number,
): Uint8Array {
  if (toEpoch < fromEpoch) {
    throw new Error(
      `Cannot derive backward: fromEpoch=${fromEpoch}, toEpoch=${toEpoch}`,
    );
  }
  if (toEpoch === fromEpoch) return key;
  let current = key;
  for (let e = fromEpoch + 1; e <= toEpoch; e++) {
    current = deriveNextEpochKey(current, spaceId, e);
  }
  return current;
}
